scapy/contrib/knx.py

```diff
@@ -0,0 +1,620 @@
+    67: "PID_TTL",
+    68: "PID_KNXNETIP_DEVICE_CAPABILITIES",
+    69: "PID_KNXNETIP_DEVICE_STATE",
+    70: "PID_KNXNETIP_ROUTING_CAPABILITIES",
+    71: "PID_PRIORITY_FIFO_ENABLED",
+    72: "PID_QUEUE_OVERFLOW_TO_IP",
+    73: "PID_QUEUE_OVERFLOW_TO_KNX",
+    74: "PID_MSG_TRANSMIT_TO_IP",
+    75: "PID_MSG_TRANSMIT_TO_KNX",
+    76: "PID_FRIENDLY_NAME",
+    78: "PID_ROUTING_BUSY_WAIT_TIME"
+}
+
+
+### KNX SPECIFIC FIELDS
+
+# KNX Standard v2.1 - 03_05_01 p.17
+class KNXAddressField(ShortField):
+    def i2repr(self, pkt, x):
+        if x is None:
+            return None
+        else:
+            return "%d.%d.%d" % ((x >> 12) & 0xf, (x >> 8) & 0xf, (x & 0xff))
+
+    def any2i(self, pkt, x):
+        if type(x) is str:
+            try:
+                a, b, c = map(int, x.split("."))
+                x = (a << 12) | (b << 8) | c
+            except:
+                raise ValueError(x)
+        return ShortField.any2i(self, pkt, x)
+
+# KNX Standard v2.1 - 03_05_01 p.18
+class KNXGroupField(ShortField):
+    def i2repr(self, pkt, x):
+        return "%d/%d/%d" % ((x >> 11) & 0x1f, (x >> 8) & 0x7, (x & 0xff))
+
+    def any2i(self, pkt, x):
+        if type(x) is str:
+            try:
+                a, b, c = map(int, x.split("/"))
+                x = (a << 11) | (b << 8) | c
+            except:
+                raise ValueError(x)
+        return ShortField.any2i(self, pkt, x)
+
+
+### KNX PLACEHOLDERS
+
+# KNX Standard v2.1 - 03_08_02 p21
+class HPAI(Packet):
+    name = "HPAI"
+    fields_desc = [
+        ByteField("structure_length", None),
+        ByteEnumField("host_protocol", 0x01, HOST_PROTOCOL_CODES),
+        IPField("ip_address", None),
+        ShortField("port", None)
+    ]
+
+    def post_build(self, p, pay):
+        p = (len(p)).to_bytes(1, byteorder='big') + p[1:]
+        return p + pay
+
+
+# DIB, KNX Standard v2.1 - 03_08_02 p22
+class ServiceFamily(Packet):
+    name = "Service Family"
+    fields_desc = [
+        ByteField("id", None),
+        ByteField("version", None)
+    ]
+
+
+# Different DIB types depends on the "description_type_code" field
+# Defining a generic DIB packet and differentiating with `dispatch_hook` or `MultipleTypeField` may better fit KNX specs
+class DIBDeviceInfo(Packet):
+    name = "DIB: DEVICE_INFO"
+    fields_desc = [
+        ByteField("structure_length", None),
+        ByteEnumField("description_type", 0x01, DESCRIPTION_TYPE_CODES),
+        ByteEnumField("knx_medium", 0x02, KNX_MEDIUM_CODES),
+        ByteField("device_status", None),
+        KNXAddressField("knx_address", None),
+        ShortField("project_installation_identifier", None),
+        XBitField("device_serial_number", None, 48),
+        IPField("device_multicast_address", None),
+        MACField("device_mac_address", None),
+        StrFixedLenField("device_friendly_name", None, 30)
+    ]
+
+    def post_build(self, p, pay):
+        p = (len(p)).to_bytes(1, byteorder='big') + p[1:]
```

Comment:
same here, and everywhere else
Also we typically use `struct.pack("!B", len(p))` which is much clearer than `to_bytes`